### PR TITLE
Store benchmark failure into JSON (issue #162)

### LINKED ISSUE #162: Validation failure not stored in JSON/CSV
When a benchmark (validation) fails, the JSON output is shorter (than in successful run) but the failure is not logged there. One has to inspect stdout for that information.

This information should be part of the JSON somehow (we already have `environment.vm.termination` field there).

Storing this inside CSV could be tricky, though.

### PR DESCRIPTION
**Pull request by vhotspur**
This is an attempt to fix #162 (JSON and CSV do not report in any way whether the benchmark failed).

In this PR the information about benchmark outcome is added to JSON. I have not added it to CSV as I see no reasonable way to store it there.

For the purposes of testing, this PR also extends the dummy benchmarks with two failing ones.

The downside is that the code for the output writers is growing in complexity. Hopefully with #81 this functionality could be moved to separate plugins.

### CODE REVIEW

File: renaissance-core/src/main/java/org/renaissance/ResultObserver.java [hunk 1]
```diff
@@ -11,5 +11,13 @@
...
   public void onNewResult(String benchmark, String metric, long value);
+
+  /** Callback when benchmark fails.
+   *
+   * Serves to e.g. invalidate previous results or marking the run as failed.
+   *
+   * @param benchmark Name of the benchmark.
+   */
+  public void onFailure(String benchmark);
   public void onExit();
```
Comment (axel22) [inline renaissance-core/src/main/java/org/renaissance/ResultObserver.java]: We should add an empty line.

File: renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java [hunk 2]
```diff
@@ -96,6 +96,9 @@ public final Throwable runBenchmark(Config config) {
       }
       return null;
     } catch (Throwable t) {
+      for (ResultObserver observer : config.resultObservers()) {
+        observer.onFailure(name());
+      }
```
Comment (axel22) [inline renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java]: What would be the consequences of placing this callback in the `finally` block after the `tearDownAfterAll` call?

File: renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java [hunk 3]
```diff
@@ -96,13 +96,19 @@ public final Throwable runBenchmark(Config config) {
...
       return t;
     } finally {
       try {
         tearDownAfterAll(config);
       } catch (Throwable t) {
         System.err.println("Error during tear-down: " + t.getMessage());
         t.printStackTrace();
+        for (ResultObserver observer : config.resultObservers()) {
+          observer.onFailure(name());
```
Comment (axel22) [inline renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java]: Another thing - should you be passing the exception object to the observer too?

Comment (axel22) [reply renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java]: Also, should `t` be returned here?

File: renaissance-core/src/main/java/org/renaissance/ResultObserver.java (hunk 1 above)
Comment (vhotspur) [reply renaissance-core/src/main/java/org/renaissance/ResultObserver.java]: Done, thanks for catching this.

File: renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java (hunk 2 above)
Comment (vhotspur) [reply renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java]: It should be there too :-). Fixed.

File: renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java (hunk 3 above)
Comment (vhotspur) [reply renaissance-core/src/main/java/org/renaissance/RenaissanceBenchmark.java]: > Also, should t be returned here?

I am not sure why we do not return `t` here but it definitely makes sense. Added.

> Another thing - should you be passing the exception object to the observer 

Maybe, it is not needed now.

**Pull request merged**